### PR TITLE
Flag pop up+code splitting dev

### PR DESCRIPTION
**Pull request by FosterBirnbaum**

### CODE REVIEW

File: homepage/frontend/src/components/questions/FlagPopUp.js [hunk 1]
```diff
@@ -1,66 +1,61 @@
...
+            >
+                {close => (
+                    <div>
+                        <a className="close" onClick={close}>
+                            &times;
+                    </a>
+                        <div className="modal-header">Flag Question</div>
+                        <div className="modal-content">
+                            <Form onSubmit={this.props.onSubmit}>
+                                <Form.Group controlId="newQuestionForm.content">
```
Comment (16chuang) [inline homepage/frontend/src/components/questions/FlagPopUp.js:41]: Tiny nitpick: update `controlId`

File: homepage/frontend/src/components/questions/QuestionCard.js [hunk 2]
```diff
@@ -144,37 +142,14 @@ export class QuestionCard extends Component {
 						<Button variant="link" onClick={this.modifyUpvotes} block><IoIosArrowDropupCircle color={this.state.upvoteColor} /></Button>
 						<p className="centered-text">{this.state.upvotes}</p>
 						<Button variant="link" onClick={this.flag} block><IoIosFlag color={this.state.flagColor} /></Button>
-						<Popup
-							position="center center"
-							modal
-							closeOnDocumentClick
-							open={this.state.flagedPopUp && !this.state.flagSubmitted}
+						<FlagPopUp
```
Comment (16chuang) [inline homepage/frontend/src/components/questions/QuestionCard.js:145]: Beautiful!

File: homepage/frontend/src/components/App.js [hunk 3]
```diff
@@ -1,40 +1,67 @@
 //Overall organization of urls. 
 
-import React, { Component, Fragment } from 'react';
+import React, { Component, Suspense, lazy } from 'react';
```
Comment (16chuang) [inline homepage/frontend/src/components/App.js:3]: I haven't been reading a lot on code splitting, so forgive the novice questions. If I understand correctly, there are two approaches: React Suspense and lazy, or loadable components? I see that we've imported both, but are we only using Suspense here? If so, should we be using `lazy` in our component imports like the example [here](https://reactjs.org/docs/code-splitting.html#reactlazy)?

```
import React, { Suspense } from 'react';

const OtherComponent = React.lazy(() => import('./OtherComponent'));

function MyComponent() {
  return (
    <div>
      <Suspense fallback={<div>Loading...</div>}>
        <OtherComponent />
      </Suspense>
    </div>
  );
}
```

File: homepage/frontend/src/components/App.js [hunk 4]
```diff
@@ -1,40 +1,67 @@
...
 import { connect } from 'react-redux';
-// import 'bootstrap/dist/css/bootstrap.min.css';
-import Header from './layout/Header';
-import QuestionPage from './questions/QuestionPage';
 import { Provider } from 'react-redux';
 import store from '../store';
 import { BrowserRouter as Router, Switch, Route } from 'react-router-dom'
+import loadable from "@loadable/component";
+
+function retry(fn, retriesLeft = 5, interval = 1000) {
```
Comment (16chuang) [inline homepage/frontend/src/components/App.js:11]: What is this retry function for?

File: homepage/frontend/src/components/App.js [hunk 5]
```diff
@@ -1,40 +1,67 @@
...
+                    }
+
+                    // Passing on "reject" is the important part
+                    retry(fn, retriesLeft - 1, interval).then(resolve, reject);
+                }, interval);
+            });
+    });
+}
+
+//Not sure why this is not working...
```
Comment (16chuang) [inline homepage/frontend/src/components/App.js:30]: How can I tell if it is working or not? (Right now the page loads immediately for me.) Should I add a delay in the functions that fetch the questions?

File: homepage/frontend/templates/frontend/index.html [hunk 6]
```diff
@@ -1,18 +1,30 @@
...
 <body>
     <div id="app"></div>
     {% load static %}
-    <script src="{% static "frontend/main.js" %}"></script>
-    <script src="https://code.jquery.com/jquery-3.4.1.slim.min.js" integrity="sha384-J6qa4849blE2+poT4WnyKhv5vZF5SrPo0iEjwBvKU7imGFAV0wwj1yYfoRSJoZ+n" crossorigin="anonymous"></script>
-    <script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js" integrity="sha384-Q6E9RHvbIyZFJoft+2mJbHaEWldlvI9IOYy5n3zV9zzTtmI3UksdQRVvoxMfooAo" crossorigin="anonymous"></script>
-    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js" integrity="sha384-wfSDF2E50Y2D1uUdj0O3uMBJnjuUD4Ih7YwaYd1iqfktj0Uod8GCExl3Og8ifwB6" crossorigin="anonymous"></script>
+    <script src="./static/frontend/About.js"></script>
+    <script src="./static/frontend/QuestionPage.js"></script>
+    <script src="./static/frontend/vendors~QuestionPage.js"></script>
```
Comment (16chuang) [inline homepage/frontend/templates/frontend/index.html:17]: What is `vendors~QuestionPage.js`/how is it different from `QuestionPage.js`?

File: homepage/frontend/templates/frontend/index.html [hunk 7]
```diff
@@ -1,18 +1,30 @@
...
 </head>
+
 <body>
     <div id="app"></div>
     {% load static %}
-    <script src="{% static "frontend/main.js" %}"></script>
-    <script src="https://code.jquery.com/jquery-3.4.1.slim.min.js" integrity="sha384-J6qa4849blE2+poT4WnyKhv5vZF5SrPo0iEjwBvKU7imGFAV0wwj1yYfoRSJoZ+n" crossorigin="anonymous"></script>
-    <script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js" integrity="sha384-Q6E9RHvbIyZFJoft+2mJbHaEWldlvI9IOYy5n3zV9zzTtmI3UksdQRVvoxMfooAo" crossorigin="anonymous"></script>
-    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js" integrity="sha384-wfSDF2E50Y2D1uUdj0O3uMBJnjuUD4Ih7YwaYd1iqfktj0Uod8GCExl3Og8ifwB6" crossorigin="anonymous"></script>
+    <script src="./static/frontend/About.js"></script>
```
Comment (16chuang) [inline homepage/frontend/templates/frontend/index.html:15]: To make sure I understand, if we create new pages (ex. login, profile), we should add them here too?

File: homepage/frontend/src/components/App.js [hunk 8]
```diff
@@ -6,57 +6,27 @@ import { connect } from 'react-redux';
...
-//    fallback: <h2>Loading</h2>
-//});
 
 import Header from './layout/Header'
-import QuestionPage from './questions/QuestionPage'
-import AboutUs from './misc/AboutUs'
-
 
-//import QuestionPage from './questions/QuestionPage'
+const AboutUs = lazy(() => import(/*webpackChunkName: 'About'  */'./misc/AboutUs'));
```
Comment (16chuang) [inline homepage/frontend/src/components/App.js:12]: Cool! How did you end end up getting this to work?

File: homepage/frontend/templates/frontend/index.html (hunk 7 above)
Comment (FosterBirnbaum) [reply homepage/frontend/templates/frontend/index.html:15]: Yes, if we want to add new pages that are in separate code bundles from the current ones. This is one of the things I want to fix later.

File: homepage/frontend/templates/frontend/index.html (hunk 6 above)
Comment (FosterBirnbaum) [reply homepage/frontend/templates/frontend/index.html:17]: Honestly, I am not entirely sure. I just know that we need to include both...

File: homepage/frontend/src/components/App.js (hunk 8 above)
Comment (FosterBirnbaum) [reply homepage/frontend/src/components/App.js:12]: I got it to work by adding the script lines in index.html, thereby ensuring the resource is present when React decides it needs to be loaded.

**Pull request merged**